### ISSUE TITLE
Does not work with multiple files

### ISSUE THREAD

**Issue by sbeyer** [wsdjeg/vim-fetch#17]
The vim invocations
```sh
vim file1:23 file2:42
vim -o file1:23 file2:42
```
only opens file1 at line 23 (in the second example: twice!) instead of file1 at line 23 and file2 at line 42.

**Comment by kopischke** [comment]
I’ll have to try to repro, as that would be a major regression (_fetch_ was specifically designed and tested to handle multiple files). Did you try this without any plugin and .vimrc customisations but _fetch_?

**Comment by sbeyer** [comment]
> Did you try this without any plugin and .vimrc customisations but fetch?

To test this, I made a .vim directory for testing, as follows:
```
~/.vim$ find . -type f | grep -v .git
./autoload/pathogen.vim
./vimrc
./bundle/vim-fetch/LICENSE.md
./bundle/vim-fetch/autoload/fetch.vim
./bundle/vim-fetch/autoload/stay/integrate/fetch.vim
./bundle/vim-fetch/doc/vim-fetch.txt
./bundle/vim-fetch/README.md
./bundle/vim-fetch/img/vim-fetch.gif
./bundle/vim-fetch/plugin/fetch.vim
~/.vim$ cat vimrc
execute pathogen#infect()
```

> fetch was specifically designed and tested to handle multiple files

I bisected the issue: 41291b30e277a46838997aa2594f3b8b94168b7e is the first commit where it cannot handle multiple files.

**Comment by kopischke** [comment]
Good news: I can repro on my side. Bad news: the culprit is not the commit you bisected but changes in Vim itself; I tested with the macOS provided Vim version (7.4.8506*), and that old-timer handles multiple files with _vim-fetch_ just fine, while my (Mac)Vim 8.0.1502 fails to do so.

If you are wondering why this is bad news: you might have noticed there have been no commits to this plugin for, golly, some years now, as RL has overtaken my dabbling in software development and my Vim usage has dwindled proportionately. This means I can’t promise I will find time to look into this in any defined timeframe, if at all, though I would merge a PR solving the issue as a matter of course.

\* EDIT: that is what Apple calls it; it’s effectively Vim 7.4.898.

**Comment by kopischke** [comment]
Not to self and other takers: #14 sounds like it might be related, if not outright the same issue.

**Comment by sbeyer** [comment]
> Bad news: the culprit is not the commit you bisected

Okay. But still: when I checkout 408a2dc (the commit before the bad one), it works fine for me (Vim 8.0 on Debian GNU/Linux).

So I reverted commit 41291b3 on current master, fixed the conflicts and now I have this commit and it works:
```
commit 78363cfe510b37e60eeb4e3c70e45f470a961b71
Author: Stephan Beyer <removed-due-to-spam@example.com>
Date:   Wed Feb 28 23:26:17 2018 +0100

    Revert "Get rid of WinEnter and TabEnter autocommands"
    
    I have no idea what I am doing but I fixed the conflicts in a way
    that it works now.

diff --git a/plugin/fetch.vim b/plugin/fetch.vim
index 63b3e94..143126b 100644
--- a/plugin/fetch.vim
+++ b/plugin/fetch.vim
@@ -9,17 +9,21 @@ endif
 let s:cpoptions = &cpoptions
 set cpoptions&vim
 
-" Based on |BufNewFile|, but flanked by |BufWinEnter| to correctly process all
-" buffers in an |arglist| passed with '-o/-O' resp. '-p' (see
-" |windows-starting| for some background, though that omits to mention that
-" |BufRead| events are also skipped, as is |BufNewFile|, that |BufWinEnter|
-" events *are* fired for all buffers, and relies on implicitly understanding
-" that the first buffer of the list is activated after loading all files, but
-" that the usual event sequence still is out of kilter; also note there is no
-" equivalent help section for '-p' though its behaviour is analogous).
+" Based on |BufWinEnter| to correctly process all buffers in the initial
+" |arglist| (see |windows-starting| for some background, though that omits to
+" mention that |BufRead| events are also skipped, as is |BufNewFile|, that
+" |BufWinEnter| events *are* fired for all buffers, and relies on implicit
+" understanding that the first buffer of the list is activated after loading,
+" thus triggering all relevant events; also note there is no equivalent help
+" section for '-p' though its behaviour is analogous).
 "
-" Note the use of highly spec specific file name patterns to avoid autocommand
-" flooding when nesting (which is needed as we switch buffers out).
+" The extra |WinEnter| and |TabEnter| events are needed to correctly process
+" the first file in an |arglist| passed with '-o/-O' resp. '-p'; without them,
+" the first buffer is correctly loaded, but its window still displays the
+" spec'ed version of the file  (go figure; no, I'm not sure I want to know).
+"
+" Note the use of spec specific file name patterns to avoid autocommand
+" flooding when nesting.
 let s:matchers = {
   \   'colon': '?*:[0123456789]*',
   \   'paren': '?*([0123456789]*)',
@@ -28,11 +32,11 @@ let s:matchers = {
   \ }
 
 " Set up autocommands:
+let s:events   = has('windows') ? 'BufWinEnter,WinEnter,TabEnter' : 'BufWinEnter,WinEnter'
 augroup fetch
   autocmd!
   for [s:spec, s:pattern] in items(s:matchers)
-    execute 'autocmd BufNewFile,BufWinEnter' s:pattern
-          \ 'nested call fetch#buffer("'.s:spec.'")'
+    execute 'autocmd' s:events s:pattern 'nested call fetch#buffer("'.s:spec.'")'
     unlet! s:spec s:pattern
   endfor
 augroup END
```

**Comment by kopischke** [comment]
I’m glad if that works for you; at the time 41291b3 was committed, the extra window event processing caused some problems for buffer and arglist handling with _fetch_ active, which is why I cut it. That might not be the case anymore, of course, considering the whole foundation seems to have subtly shifted…

**Comment by kopischke** [comment]
BTW, only the failure to handle the split window call is actually a bug; the fact that your `vim file1:23 file2:42` call only opens `file1` is Vim working as designed. As it happens, your `file2` is on the buffer list (still adorned with its column indicator) and will be processed by _vim-fetch_ when actually loaded into a window. The first item in the troubleshooting section of _fetch_’s help mentions that, though I admit that its referring the arglist may have been muddying the waters.

**Comment by sbeyer** [comment]
Good that you say that. When I use the `:n[ext]` command when file1:23 is open, it says "Cannot go beyond last file". It did so before my "fix" and it does so after the fix. (*edit*: However, file1:23 and file2:42 are in the buffers list, so that is fine. Just in the wrong direction, I have to use `:p[rev]`. This is before and after my fix.) So as you say, my fix only works for the split window case.

I have by the way no idea what the code is doing, my conflict fixing was just blind trial-and-error.

(*edit2, totally unrelated to the issue*: I btw really wonder why it is so hard in vim to get a seemingly simple feature like that. I already switched from the `file-line` vim script to vim-fetch because `file-line` was so extremely buggy.)

**Comment by kopischke** [comment]
It will work fine if you use buffer commands (i.e. `:bn[ext]`) instead of arglist commands; those are failing because _fetch_’s arglist futzing does not work as expected: it should insert the resolved file at the index of the original spec, but it seems to append it instead (which would be the issue #14 is about, now that I think about it).

EDIT (note to self): futzing appends a spurious duplicate entry to the arglist because only the part of `:h argedit` is true that states that the command is equivalent to `:argadd` followed by `:edit`, which means the entry is always added. EDIT2EDIT: futzing works as originally designed in Vim 7.4.858.

**Comment by kopischke** [comment]
>(edit2, totally unrelated to the issue: I btw really wonder why it is so hard in vim to get a seemingly simple feature like that. I already switched from the file-line vim script to vim-fetch because file-line was so extremely buggy.)

Honestly? Because Vim is a decade old beast, a crusty steam engine made of technical debt with extension points haphazardly bolted onto it. Its core editing model (composable syntax, text objects, ex commands) is f****tastic, but the technical platform is painful to work with.

**Comment by kopischke** [comment]
Because I  can’t let go of stuff (and, accessorily, because the flu stops me from doing more productive work) I have been testing permutations of autocommands and multiple file edits. The bad news is: reverting  41291b3 only helps for a combination of two files; try three or more and things still fail. The worse news is that no other combination I can think of works (despite the fact that, adding assault to injury, `BufNewFile` is triggered for all files in all permutations – just not when _fetch_ tries to process it).

**Comment by sbeyer** [comment]
> The bad news is: reverting 41291b3 only helps for a combination of two files; try three or more and things still fail.

You are right. This is so f**ked up.

**Comment by kopischke** [comment]
Yeah, apparently switching a buffer out stops any further processing of the triggering autocommand; the revert alleviates the situation because it makes _fetch_ trigger on different autocommands for the first two files, but after that, you are, as you aptly stated, f***ed.

**Comment by kopischke** [comment]
Heretical, but not entirely unrelated, thought: have you had a look a Atom lately? I find myself liking it more and more, not least because their dev team is willing to think about APIs and does not shy from breaking backwards compatibility; they have essentially obviated most of their package ecosystem in the last six months, and yes, that is a good thing, think [Language Server Protocol](https://microsoft.github.io/language-server-protocol//language-server-protocol/) and [Tree Sitter](https://github.com/tree-sitter/tree-sitter) grammars.

EDIT: need I mention that `file:line` and `file:line:col` work out of the box?

**Comment by sbeyer** [comment]
I once looked at it because it was hyped in some way but (1) I want an editor that also works in a terminal because I also want to code in ssh sessions, and (2) I want an editor that does not eat half of my CPU/memory. Vim's IDE capabilities are limited, though, so I am willing to check other editors but I doubt that Atom will be my choice. Maybe I should switch to Emacs :dagger:

**Comment by kopischke** [comment]
I see where you are coming from; I’ve always been a GUI Vim user myself, so Atom seems less of a jump, especially now that their startup speed has very much improved and resource usage is down to something bearable.

More to the point, I have good news: I think I have found a way to make _fetch_ work reliably both on Vim 7.4 (or even older versions of Vim 7, but I can’t test these) and Vim 8. I would ask you to take the new version around the block as soon as I push it, if that is OK with you?

**Comment by kopischke** [comment]
> Maybe I should switch to Emacs 🗡

Now _that_ is a heretical thought 😝 .

**Comment by kopischke** [comment]
@sbeyer could you please test the `release-3.0.0` branch?

**Comment by sbeyer** [comment]
Thank you @kopischke.

I've tested `vim file1:23 file2:42 file3:5 file2:12 file3`. It worked with `:n` and `:bn` as expected. One minor thing: the last buffer (file3) was also at line 5. If I exchange `file3:5` and `file3`, everything works as expected.

When adding `-o` or `-O`, everything works as expected.

I've also tested this using lines that do not exist and it works as expected.

Thanks, great job!

*edit, for the record*: Only tested with vim 8.0

**Comment by kopischke** [comment]
> One minor thing: the last buffer (file3) was also at line 5. If I exchange `file3:5` and `file3`, everything works as expected.

Huh. Must check that, it sound like it could get annoying fast. Thanks for taking the branch for a whirl!

**Comment by sbeyer** [comment]
> Must check that, it sound like it could get annoying fast

If it's not easy to fix, keep it that way. That was an artificial test example. In reality, I'd always have either all files with line information or all files without line information. (Also there is no problem with `-o`.)

**Comment by kopischke** [comment]
Could you post the output of `:args` when that happens here? I have an inkling what happens and might be able to eliminate that glitch if I am right.

**Comment by sbeyer** [comment]
Output of `:args`: `[file1] file2:42 file3:5 file2:12 file3`

Another very minor issue is that vim says: `"file1:23" [New File]` when opening the file. (But it's not a new file.) It's just the output that is wrong, saving and re-loading works without problems.

**Comment by kopischke** [comment]
Ah, I should have been more specific: I need the output of `:args` **after** you have switched to `file3:5`, and, for good measure, the same thing after switching a second time, to `file3`. Also please do tell how exactly you switch (`:next`?).

>Another very minor issue is that vim says: `"file1:23" [New File]` when opening the file.

When exactly does it say that? Technically, that **is** a new file  `$pwd/file1:23`, which _fetch_ then swaps out, but I can see how that message could be confusing. Could you post the output of `:echo &shortmess`, please?

**Comment by sbeyer** [comment]
> I need the output of :args after you have switched to file3:5

`file1 file2 [file3] file2:12 file3`

> for good measure, the same thing after switching a second time, to file3

`file1 file2 file3 file2 [file3]`

> Also please do tell how exactly you switch (:next?).

The output above comes from `:next`. When I use `:bn`, the output at `file3:5` is the same, but when I do the next `:bn` things get weird:
* The next `:bn` goes to `file1` again but `:args` still shows `file1 file2 [file3] file2:12 file3`
* the next `:bn` goes to `file2:42` again but `:args` still shows `file1 file2 [file3] file2:12 file3`
* the next `:bn` goes to `file2:12` and `:args` shows `file1 file2 file3 [file2] file3`
* the next `:bn` goes to `file3` (but on line 5) and `:args` shows `file1 file2 file3 [file2] file3`
* the next `:bn` goes to `file1` and `:args` still shows `file1 file2 file3 [file2] file3`
* every further `:bn` cycles further but `:args` still shows `file1 file2 file3 [file2] file3`

With `:buffers` I get
```
:buffers
  2      "file2:42"                     line 0
  3      "file3:5"                      line 0
  4      "file2:12"                     line 0
  5      "file3"                        line 0
  6 %a   "file1"                        line 23
:bn
:buffers
  3      "file3:5"                      line 0
  4      "file2:12"                     line 0
  5      "file3"                        line 0
  6 #    "file1"                        line 23
  7 %a   "file2"                        line 42
:bn
:buffers
  4      "file2:12"                     line 0
  5 %a   "file3"                        line 5
  6      "file1"                        line 23
  7 #    "file2"                        line 42
:bn
:buffers
  4      "file2:12"                     line 0
  5 #    "file3"                        line 5
  6 %a   "file1"                        line 23
  7      "file2"                        line 42
:bn
:buffers
  4      "file2:12"                     line 0
  5      "file3"                        line 5
  6 #    "file1"                        line 23
  7 %a   "file2"                        line 42
...
```

And using `:n`:
```
:buffers
  2      "file2:42"                     line 0
  3      "file3:5"                      line 0
  4      "file2:12"                     line 0
  5      "file3"                        line 0
  6 %a   "file1"                        line 23
:n
:buffers
  3      "file3:5"                      line 0
  4      "file2:12"                     line 0
  5      "file3"                        line 0
  6 #    "file1"                        line 23
  7 %a   "file2"                        line 42
:n
:buffers
  4      "file2:12"                     line 0
  5 %a   "file3"                        line 5
  6      "file1"                        line 23
  7 #    "file2"                        line 42
:n
:buffers
  5 #    "file3"                        line 5
  6      "file1"                        line 23
  7 %a   "file2"                        line 12
:n
:buffers
  5 %a   "file3"                        line 5
  6      "file1"                        line 23
  7 #    "file2"                        line 12
:n -> E165: Cannot go beyond last file
```

> > Another very minor issue is that vim says: "file1:23" [New File] when opening the file.
>
> When exactly does it say that?

When it opens the file, e.g., directly after invoking vim, after `:n`, and after `:bn`.

For a comparison, an invocation without line numbers (or using `:e`) says `"file1" 115L, 307C`.

Sorry, that I'm afk most of the time. If you need more prompt help with hunting down the bug, we should meet on some real-time messenger or something :)

**Comment by kopischke** [comment]
That seemingly odd behavior of the buffer list is expected: when _fetch_ swaps out the spurious `name:XXX` buffers, some of these turn out to be identical (i.e. `file3:5` and `file3` are both `file3`), which means they will only appear once in your buffer list (which, unlike the argument list, cannot contain duplicates: a buffer is either atomically open or absent from the list). That also explains why “all” `file3` buffers are on the same line when the swap has happened: there is only one `file3` buffer, which you switch to. The same thing happens on the argument list, but that is _fetch_ removing the spurious entries to keep things tidy.

Things work more like you seem to expect them to when you split windows (i.e. with the `-o` and `-O` switches, or with the split prefixes to editing commands), as one and the same buffer can scroll to different positions in different windows.

>>>Another very minor issue is that vim says: "file1:23" [New File] when opening the file.
>>
>>When exactly does it say that?
>
>When it opens the file, e.g., directly after invoking vim, after `:n`, and after `:bn`. For a comparison, an invocation without line numbers (or using :e) says `"file1" 115L, 307C`.

I suspect you have removed `o`, `O` or both from your `shortmess`, which would explain that phenomenon: as I mentioned, there really is a new file, which _fetch_ swaps out immediately. With those options missing, the message for reading the swapped out file does not overwrite the message for reading the spurious first one.

I’ll wait for you posting the `shortness` value before considering this issue solved, but all this sounds as if things are actually working as expected, given expectations familiar with the inner workings of both Vim and _fetch_ 😉 .

**Comment by kopischke** [comment]
>Sorry, that I'm afk most of the time. If you need more prompt help with hunting down the bug, we should meet on some real-time messenger or something :)

No worries, you have ben amazingly responsive; I just happen to have a lazy Sunday on my hands 🤓 .

**Comment by sbeyer** [comment]
> I suspect you have removed o, O or both from your shortmess, which would explain that phenomenon

`shortmess=filnxtToOc`

**Comment by kopischke** [comment]
>`shortmess=filnxtToOc`

Huh. Hmmm. Oh, I think I know where that comes from…

**Comment by kopischke** [comment]
@sbeyer I pushed some fixup commits meant to restore normal message functionality (the `"file1" 115L, 307C` message). Could you test if the current branch head works as expected? Just pull the `release-3.0.0` branch again.

**Comment by sbeyer** [comment]
Now there is a short moment where it shows the "[New file]" thing (I guess vim adds a short delay for that message) and then everything is fine. It only does that for the first file, so the delay is not an issue.

**Comment by kopischke** [comment]
I didn’t notice the “new file” message myself, but that might be a latency issue. As I said before, there actually is a new file, it’s just that in normal operation mode, its loading message will be overwritten when _fetch_ swaps the buffer out…

I will file this under “all is well”, then; thanks a lot for your patience and helpfulness!

**Comment by sbeyer** [comment]
Thanks a lot to *you* for fixing the bug and make using vim a pleasant experience (although you do not use vim yourself any longer). ;)

**Comment by kopischke** [comment]
Closed by ef8a8e78f9bb724ca0fd2bd0b4e39d3e12487832.
